DOCSP-30724: Connection Troubleshooting page (#115)
mongodb/docs-kotlin: source/connection-troubleshooting.txt

~~~~~~~~~~~~~~~~~~~~

The {+driver-short+} can fail to connect to a MongoDB instance if
the authorization is not configured correctly. This often results in an error
message similar to the following:

.. code-block:: none
   :copyable: false

   Command failed with error 18 (AuthenticationFailed): 'Authentication failed.' on server localhost:27017.

If you receive this error, try the following methods to resolve the issue.

.. _kotlin-connection-string-auth:

Check Connection String
-----------------------

An invalid connection string is the most common cause of authentication
issues when attempting to connect to MongoDB.

.. note::

   For more information about using connection strings with the {+driver-short+},
   see :ref:`Connection URI <connection-uri>` in the Connection Guide.

If your connection string contains a username and password, ensure that they
are in the correct format.

.. note::

   If the username or password includes any of the following characters, they
   must be `percent encoded <https://tools.ietf.org/html/rfc3986#section-2.1>`__:

   .. code-block:: none

      : / ? # [ ] @

If your MongoDB deployment is on MongoDB Atlas, you can check your connection

> string by using the :ref:`Connect to MongoDB Atlas <connect-atlas-kotlin-driver>`
> code example. Make sure to replace the connection string in the example.

Reviewer: The linked example has a different title than the one in the line

```suggestion
string by using the :ref:`Atlas Connection Example <connect-atlas-kotlin-driver>`.
Make sure to replace the connection string in the example with yours.
```

When connecting to a replica set, you should include all of the hosts
in the replica set in your connection string. Separate each of the hosts
in the connection string with a comma. This enables the driver to establish a
connection if one of the hosts is unreachable.

.. _kotlin-connection-admin:

Verify User Is in Authentication Database
-----------------------------------------

To successfully authenticate a connection by using a username and password,
the username must be defined in the authentication database. The default
authentication database is the ``admin`` database. To use a different database
for authentication, specify the ``authSource`` in the connection string. The
following example instructs the driver to use ``users`` as the authentication
database:

.. code-block:: kotlin
   :copyable: false

>    val mongoClient =
>    MongoClients.create("mongodb://<username>:<password>@<hostname>:<port>/?authSource=users");

Reviewer: I: make sure that your indentation doesn't affect your code examples. Also, I don't think this code example is correct, per the connection guide

```suggestion
   val mongoClient = MongoClient.create("mongodb://<username>:<password>@<hostname>:<port>/?authSource=users")
```

   
.. _kotlin-error-sending-message:

Error Sending Message
~~~~~~~~~~~~~~~~~~~~~

> When you end a request through the driver and it is unable to send the command,

Reviewer: ```suggestion
When you send a request through the driver and it is unable to send the command,
```
